### ISSUE TITLE
What does it mean to "train the curve"?

### ISSUE THREAD

**Issue by Zephyrasgit** [timgaripov/dnn-mode-connectivity#5]
From my understanding of the paper, we are finding areas of low loss between two distinct checkpoints of the same neural network. This should mean that we are finding the curve between this two checkpoints.

What does it mean to "train the curve" from the two checkpoints, why does it give us new checkpoints, and what does it mean to "evaluate the curve" when we input in only 1 checkpoint in eval_curve.py?

**Comment by izmailovpavel** [comment]
Hey @Zephyrasgit, thanks for your question. The curve is parametric and in fact has as many parameters as the neural network architecture that you are using. Training means that we optimize the parameters of the curve. You can think of the training procedure as follows: imagine 2 points in a 3D space. You want to connect them with a polygonal chain with 2 segments. Then, you have to pick the position of the bend of the curve, a point in 3D space. That's exactly what we are doing, but instead of the 3D space we have the ~million-dimensional space of neural network parameters.

The checkpoint that the method outputs contains the parameters of the curve as well as the information about the endpoints of the curve. That's why you have just one checkpoint as the output / input to the eval script. 

Please let us know if you have any other questions.

**Comment by Zephyrasgit** [comment]
Thanks for the quick response and clarification, it helped me understand much more about the code!

Just some last questions: you say that the checkpoint from training a curve outputs a checkpoint that contains parameters of the curve and endpoints of the curve. Can I know how this is stored within one checkpoint? I am guessing it has to do with the CurveNet and individual curve format of the models, something to do with fix_points, is that correct?

**Comment by izmailovpavel** [comment]
Hi @Zephyrasgit, you are correct, it has to do with the curve class (see [here](https://github.com/timgaripov/dnn-mode-connectivity/blob/master/curves.py)). We re-define the architecture of the network so that a single instance represents the whole curve. In particular, every layer stores the parameters of each bend of the curve and the endpoints. For example, for a polygonal chain with one bend, each layer would store three sets of parameters.

You can set the endpoints by using [this method](https://github.com/timgaripov/dnn-mode-connectivity/blob/77b67db68f3fe20eefddd0ec73275250221ac61f/curves.py#L274-L278) or export them using [this method](https://github.com/timgaripov/dnn-mode-connectivity/blob/77b67db68f3fe20eefddd0ec73275250221ac61f/curves.py#L284-L288). See how it's done here: we load the endpoints of the curve from the checkpoints in our [curve training script](https://github.com/timgaripov/dnn-mode-connectivity/blob/master/train.py#L103-L110).
